pincer/core/gateway.py

```diff
@@ -310,19 +311,23 @@ async def handle_reconnect(self, payload: GatewayDispatch):
             self.shard_key
         )
 
+        self.__should_resume = True
         await self.__socket.close()
-        self.__should_reconnect = True
-        await self.start_loop()
 
     async def handle_invalid_session(self, payload: GatewayDispatch):
         """|coro|
         Opcode 9 - Invalid connection
         Attempt to relog. This is probably because the session was already invalidated
         when we tried to reconnect.
         """
-        _log.debug("%s Invalid session, attempting to relog...", self.shard_key)
-        self.__should_reconnect = False
-        await self.start_loop()
+
+        _log.debug(
+            "%s Invalid session, attempting to %s...",
+            self.shard_key,
+            "reconnect" if payload.data else "relog"
+        )
+
+        self.__should_resume = payload.data
```

Comment:
If this is typed as a Boolean we should keep it as such.
```suggestion
        self.__should_resume = bool(payload.data)
```

Reply:
It is a bool. I will update the typehint to reflect that.